### PR TITLE
ui: offline mode 

### PR DESCRIPTION
**Pull request by Rolph1**
### **In short:**
This PR significantly enhances the user interface and offline mode screen logic to have an overall better user experience. It finalizes the offline mode by adding dynamic screen changes along with definitive tests.

### **Files modified:**
-`OfflineInterviewModule`, `OfflineInterviewModuleTest`: Fixed the colors to match the theme and added a new box containing what part of the interview the interviewer should focus on.

-`OfflineRecordingsProfile`: is now cleaner and adapts to dark mode

-`PreviousRecordingsFeedback`,  `PreviousRecordingsFeedbackTest`: is now cleaner, adapts to dark mode, contains what the user said and updates as soon as the GPT response is received by polling the cache.
WhatsApp Image 2024-12-20 at 01 34 03
WhatsApp Image 2024-12-20 at 01 34 13

-`OfflinePromptsFunctions`: Are now adapted to all the other screens' logic needs

-`RecordingReviewScreen`: Now adapts to dark mode

- `LeaderBoard`: there was a typo in the parameters

- `SpeakingViewModel`,  `ChatViewModel`: the prompts are now adapted to ask GPT to mention the part of the interview to focus on
- `OfflineRecordingScreenTest`,`OfflineRecordingScreen`

-, `MainActivity`, `Theme`: A few themes have been added for the previous interview prompt cards' needs. MainActivity has been changed to adapt to the new parameters too.
### **Files deleted:**
-`OfflinePracticeQuestionsScreen`, `OfflinePracticeQuestionsScreenTest`: These files are now redundant because of the outlined text box in `OfflineInterviewModule` which now include the question.

### **In conclusion:**
This PR heavily augments the offline mode to be a fully fledged finalized feature that works very well and is reactive, a lot of bugs and inconsistencies have been fixed too. Tests are not all that thorough mostly because of multithreading to poll the cache to update the feedback screen as soon as a response has been added.

### CODE REVIEW

File: app/src/main/java/com/github/se/orator/ui/offline/OfflineInterviewModule.kt [hunk 1]
```diff
@@ -56,78 +61,124 @@ fun OfflineInterviewModule(
...
+                          textColor = MaterialTheme.colorScheme.onSurface,
+                          focusedBorderColor = MaterialTheme.colorScheme.outline,
+                          unfocusedBorderColor = MaterialTheme.colorScheme.outlineVariant,
+                          cursorColor = MaterialTheme.colorScheme.primary,
+                          focusedLabelColor = MaterialTheme.colorScheme.onSurface,
+                          unfocusedLabelColor = MaterialTheme.colorScheme.onSurfaceVariant))
+
+              Spacer(modifier = Modifier.height(AppDimensions.paddingLarge))
+
+              Button(
```
Comment (helkhour) [inline app/src/main/java/com/github/se/orator/ui/offline/OfflineInterviewModule.kt:150]: Starting from the 2nd trial, this button keeps crashing the app.

Comment (MathisHage) [reply app/src/main/java/com/github/se/orator/ui/offline/OfflineInterviewModule.kt:150]: Indeed. Here is the error message : 
java.lang.IllegalArgumentException: Navigation destination that matches request NavDeepLinkRequest{ uri=android-app://androidx.navigation/offline_recording/ } cannot be found in the navigation graph ComposeNavGraph(0x0) startDestination={ComposeNavGraph(0x78c6a2d5) route=Auth startDestination={Destination(0x594b8837) route=Auth Screen}}

File: app/src/main/java/com/github/se/orator/ui/offline/OfflineInterviewModule.kt [hunk 2]
```diff
@@ -56,78 +61,124 @@ fun OfflineInterviewModule(
 ) {
   var targetCompany by remember { mutableStateOf("") }
   var jobPosition by remember { mutableStateOf("") }
+  var question by remember { mutableStateOf("") }
   val ID = generateRandomString()
```
Comment (helkhour) [inline app/src/main/java/com/github/se/orator/ui/offline/OfflineInterviewModule.kt:66]:
```suggestion
    var lastQuestion by remember { mutableStateOf<String?>(null) } // Local state for the last question
```

File: app/src/main/java/com/github/se/orator/ui/offline/OfflineInterviewModule.kt [hunk 3]
```diff
@@ -56,78 +61,124 @@ fun OfflineInterviewModule(
...
+                    navigationActions.goToOfflineRecording(question)
+                  },
+                  modifier =
+                      Modifier.fillMaxWidth(0.8f)
+                          .padding(AppDimensions.paddingSmall)
+                          .testTag("doneButton")
+                          .align(Alignment.CenterHorizontally),
+                  colors =
+                      ButtonDefaults.buttonColors(
+                          containerColor = MaterialTheme.colorScheme.primary)) {
```
Comment (helkhour) [inline app/src/main/java/com/github/se/orator/ui/offline/OfflineInterviewModule.kt:176]:
```suggestion
                          containerColor = MaterialTheme.colorScheme.primary),
                           enabled = question.isNotEmpty() || !lastQuestion.isNullOrEmpty(), // Disable if no valid question     
                          ) {
```

File: app/src/main/java/com/github/se/orator/ui/offline/OfflineInterviewModule.kt [hunk 4]
```diff
@@ -56,78 +61,124 @@ fun OfflineInterviewModule(
...
+                                "jobPosition" to jobPosition,
+                                "question" to question,
+                                "ID" to ID,
+                                "transcribed" to "0",
+                                "GPTresponse" to "0",
+                                "transcription" to ""),
+                    )
+                    offlinePromptsFunctions.createEmptyPromptFile(context, ID)
+                    speakingViewModel.interviewPromptNb.value = ID
+                    navigationActions.goToOfflineRecording(question)
```
Comment (helkhour) [inline app/src/main/java/com/github/se/orator/ui/offline/OfflineInterviewModule.kt:167]:
```suggestion                     
                     navigationActions.goToOfflineRecording(question.ifEmpty { lastQuestion ?: "" })

```

File: app/src/main/java/com/github/se/orator/ui/offline/OfflineInterviewModule.kt (hunk 1 above)
Comment (helkhour) [reply app/src/main/java/com/github/se/orator/ui/offline/OfflineInterviewModule.kt:150]: You can try now, it should be fixed.

File: app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt [hunk 5]
```diff
@@ -266,11 +266,12 @@ class ChatViewModel(
     viewModelScope.launch {
       try {
         _isLoading.value = true
-
+        val question = offlinePromptsFunctions.getPromptMapElement(interviewID, "question", context)
```
Comment (helkhour) [inline app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt:269]: ok

**Review by helkhour — CHANGES_REQUESTED** [review]
Overall, great work!

I ran your code and noticed a small bug: the "Go to Offline Recording" button crashes the app when no previous question is filled in. Upon investigating the code, I found the issue was due to navigating to a screen without providing the required question input. I made a quick fix and committed the changes—please take a look when you get a chance.

The rest of the code looks solid! I really appreciate how you've modularized your implementation—most methods are now reusable throughout the codebase, which is fantastic.

It would have been great to reuse the speaking screens from the offline mode as discussed in previous PRs, but I understand the distinctive offline mode approach you went with here—it works well in this context.

One concern: I noticed the code coverage is quite low. Is this an issue on your end, or could it be a problem with SonarCloud?

Looking forward to your thoughts!

File: app/src/main/java/com/github/se/orator/ui/offline/OfflineInterviewModule.kt [hunk 6]
```diff
@@ -56,78 +61,124 @@ fun OfflineInterviewModule(
...
+                  singleLine = false,
+                  colors =
+                      androidx.compose.material.TextFieldDefaults.outlinedTextFieldColors(
+                          backgroundColor = MaterialTheme.colorScheme.surface,
+                          textColor = MaterialTheme.colorScheme.onSurface,
+                          focusedBorderColor = MaterialTheme.colorScheme.outline,
+                          unfocusedBorderColor = MaterialTheme.colorScheme.outlineVariant,
+                          cursorColor = MaterialTheme.colorScheme.primary,
+                          focusedLabelColor = MaterialTheme.colorScheme.onSurface,
+                          unfocusedLabelColor = MaterialTheme.colorScheme.onSurfaceVariant))
```
[the remark this answers is not in the document]
Comment (Rolph1) [reply app/src/main/java/com/github/se/orator/ui/offline/OfflineInterviewModule.kt:146]: Done. I will commit it now.

File: app/src/main/java/com/github/se/orator/ui/offline/RecordingReviewScreen.kt
[the remark this answers is not in the document]
Comment (Rolph1) [reply app/src/main/java/com/github/se/orator/ui/offline/RecordingReviewScreen.kt:1]: Done.

**Comment by Rolph1** [comment]
This PR started glitching with git especially the end to end test file. Another one has been made.
